File: alert_manager.py
import threading
import time
import subprocess
from typing import Dict, Optional
import os
import json

try:
    import winsound
    WINSOUND_AVAILABLE = True
except ImportError:
    WINSOUND_AVAILABLE = False

class AlertManager:
    def __init__(self):
        self.active_alerts: Dict[str, threading.Thread] = {}
        self.alert_stop_flags: Dict[str, bool] = {}
        self.settings_file = "audio_settings.json"
        
        # Default settings
        self.alert_timeout = 120  # 2 minutes in seconds
        self.volume = 100  # Volume percentage (1-100)
        self.beep_frequency = 880  # Hz
        self.beep_duration = 500  # milliseconds
        self.beep_interval = 1.0  # seconds between beeps
        
        # Load saved settings
        self.load_settings()

    def _play_alert(self, timer_name: str):
        """Play a beep sound repeatedly until stopped"""
        while not self.alert_stop_flags.get(timer_name, True):
            try:
                if WINSOUND_AVAILABLE:
                    # Use Windows beep with configurable settings
                    winsound.Beep(self.beep_frequency, self.beep_duration)
                else:
                    # Fallback for non-Windows systems
                    print(f"\a")  # ASCII bell
                time.sleep(self.beep_interval)
            except Exception as e:
                print(f"Error playing alert: {str(e)}")
                # If sound fails, try a simple print
                print(f"\a\nTimer {timer_name} completed!")  # \a is ASCII bell
                break

    def start_alert(self, timer_name: str):
        """Start a new alert for a timer"""
        # Stop existing alert if any
        self.stop_alert(timer_name)

        # Set up new alert
        self.alert_stop_flags[timer_name] = False
        alert_thread = threading.Thread(
            target=self._alert_with_timeout,
            args=(timer_name,)
        )
        alert_thread.daemon = True
        self.active_alerts[timer_name] = alert_thread
        alert_thread.start()

    def _alert_with_timeout(self, timer_name: str):
        """Run alert with 2-minute timeout"""
        start_time = time.time()
        self._play_alert(timer_name)

        # Stop if 2 minutes have passed
        if time.time() - start_time >= self.alert_timeout:
            self.stop_alert(timer_name)

    def stop_alert(self, timer_name: str):
        """Stop an active alert"""
        if timer_name in self.alert_stop_flags:
            self.alert_stop_flags[timer_name] = True
            if timer_name in self.active_alerts:
                self.active_alerts[timer_name].join(0.1)
                del self.active_alerts[timer_name]
            del self.alert_stop_flags[timer_name]

    def stop_all_alerts(self):
        """Stop all active alerts"""
        for timer_name in list(self.active_alerts.keys()):
            self.stop_alert(timer_name)
    
    def set_audio_settings(self, frequency=None, duration=None, interval=None):
        """Update audio settings for notifications"""
        if frequency is not None:
            self.beep_frequency = max(37, min(32767, int(frequency)))  # Windows beep limits
        if duration is not None:
            self.beep_duration = max(10, min(5000, int(duration)))  # Reasonable duration limits
        if interval is not None:
            self.beep_interval = max(0.1, min(10.0, float(interval)))  # Reasonable interval limits
        
        # Automatically save settings when changed
        self.save_settings()
    
    def get_audio_settings(self):
        """Get current audio settings"""
        return {
            'frequency': self.beep_frequency,
            'duration': self.beep_duration,
            'interval': self.beep_interval
        }
    
    def save_settings(self):
        """Save current audio settings to file"""
        settings = {
            'frequency': self.beep_frequency,
            'duration': self.beep_duration,
            'interval': self.beep_interval,
            'alert_timeout': self.alert_timeout
        }
        try:
            with open(self.settings_file, 'w') as f:
                json.dump(settings, f, indent=2)
        except Exception as e:
            print(f"Error saving audio settings: {str(e)}")
    
    def load_settings(self):
        """Load audio settings from file"""
        if not os.path.exists(self.settings_file):
            return
        try:
            with open(self.settings_file, 'r') as f:
                settings = json.load(f)
                self.beep_frequency = settings.get('frequency', 880)
                self.beep_duration = settings.get('duration', 500)
                self.beep_interval = settings.get('interval', 1.0)
                self.alert_timeout = settings.get('alert_timeout', 120)
        except Exception as e:
            print(f"Error loading audio settings: {str(e)}")

File: src/js/alert-manager.js
class AlertManager {
    constructor() {
        this.activeAlerts = new Map();
        this.alertStopFlags = new Map();
        this.settingsFile = "audio_settings.json";
        
        // Default settings
        this.alertTimeout = 120; // 2 minutes in seconds
        this.volume = 100; // Volume percentage (1-100)
        this.beepFrequency = 880; // Hz
        this.beepDuration = 500; // milliseconds
        this.beepInterval = 1.0; // seconds between beeps
        
        // Audio context for cross-platform beep generation
        this.audioContext = null;
        this.initializeAudioContext();
        
        // Load saved settings
        this.loadSettings();
    }

    initializeAudioContext() {
        try {
            // Initialize Web Audio API
            this.audioContext = new (window.AudioContext || window.webkitAudioContext)();
        } catch (e) {
            console.warn('Web Audio API not supported, using fallback beep');
            this.audioContext = null;
        }
    }

    _playBeep() {
        if (this.audioContext) {
            try {
                // Create oscillator for beep sound
                const oscillator = this.audioContext.createOscillator();
                const gainNode = this.audioContext.createGain();
                
                oscillator.connect(gainNode);
                gainNode.connect(this.audioContext.destination);
                
                oscillator.frequency.setValueAtTime(this.beepFrequency, this.audioContext.currentTime);
                oscillator.type = 'square';
                
                // Set volume (convert percentage to gain)
                gainNode.gain.setValueAtTime(this.volume / 100 * 0.1, this.audioContext.currentTime);
                
                // Play beep for specified duration
                oscillator.start(this.audioContext.currentTime);
                oscillator.stop(this.audioContext.currentTime + this.beepDuration / 1000);
                
                return true;
            } catch (e) {
                console.warn('Failed to play beep with Web Audio API:', e);
                return false;
            }
        }
        return false;
    }

    _playAlert(timerName) {
        const playLoop = () => {
            if (this.alertStopFlags.get(timerName)) {
                return; // Stop the loop
            }

            // Try to play beep, fallback to console beep if failed
            if (!this._playBeep()) {
                console.log('\x07'); // ASCII bell character
                console.log(`Timer ${timerName} completed!`);
            }

            // Schedule next beep
            setTimeout(playLoop, this.beepInterval * 1000);
        };

        playLoop();
    }

    startAlert(timerName) {
        // Stop existing alert if any
        this.stopAlert(timerName);

        // Resume audio context if suspended (required for some browsers)
        if (this.audioContext && this.audioContext.state === 'suspended') {
            this.audioContext.resume();
        }

        // Show desktop notification
        this._showDesktopNotification(timerName);

        // Set up new alert
        this.alertStopFlags.set(timerName, false);
        
        // Start alert with timeout
        this._alertWithTimeout(timerName);
    }

    _alertWithTimeout(timerName) {
        const startTime = Date.now();
        
        // Start the alert loop
        this._playAlert(timerName);
        
        // Set timeout to stop alert after specified duration
        const timeoutId = setTimeout(() => {
            this.stopAlert(timerName);
        }, this.alertTimeout * 1000);
        
        this.activeAlerts.set(timerName, timeoutId);
    }

    stopAlert(timerName) {
        if (this.alertStopFlags.has(timerName)) {
            this.alertStopFlags.set(timerName, true);
            
            if (this.activeAlerts.has(timerName)) {
                clearTimeout(this.activeAlerts.get(timerName));
                this.activeAlerts.delete(timerName);
            }
            
            // Clean up after a short delay
            setTimeout(() => {
                this.alertStopFlags.delete(timerName);
            }, 1000);
        }
    }

    stopAllAlerts() {
        for (const timerName of this.activeAlerts.keys()) {
            this.stopAlert(timerName);
        }
    }

    setAudioSettings(frequency = null, duration = null, interval = null) {
        if (frequency !== null) {
            this.beepFrequency = Math.max(37, Math.min(20000, parseInt(frequency)));
        }
        if (duration !== null) {
            this.beepDuration = Math.max(10, Math.min(5000, parseInt(duration)));
        }
        if (interval !== null) {
            this.beepInterval = Math.max(0.1, Math.min(10.0, parseFloat(interval)));
        }
        
        // Automatically save settings when changed
        this.saveSettings();
    }

    getAudioSettings() {
        return {
            frequency: this.beepFrequency,
            duration: this.beepDuration,
            interval: this.beepInterval
        };
    }

    async saveSettings() {
        const settings = {
            frequency: this.beepFrequency,
            duration: this.beepDuration,
            interval: this.beepInterval,
            alert_timeout: this.alertTimeout
        };
        
        try {
            // Use file-based persistence through Electron IPC
            if (window.electronAPI) {
                await window.electronAPI.saveFile(this.settingsFile, settings);
            }
        } catch (error) {
            console.error('Error saving audio settings:', error);
        }
    }

    async loadSettings() {
        try {
            // Use file-based persistence through Electron IPC
            if (window.electronAPI) {
                const result = await window.electronAPI.loadFile(this.settingsFile);
                if (result.success && result.data) {
                    const settings = result.data;
                    this.beepFrequency = settings.frequency || 880;
                    this.beepDuration = settings.duration || 500;
                    this.beepInterval = settings.interval || 1.0;
                    this.alertTimeout = settings.alert_timeout || 120;
                }
            }
        } catch (error) {
            console.error('Error loading audio settings:', error);
        }
    }

    // Test beep function for settings
    testBeep() {
        if (!this._playBeep()) {
            console.log('\x07'); // Fallback beep
            console.log('Test beep!');
        }
    }

    // Show desktop notification
    async _showDesktopNotification(timerName) {
        try {
            if (window.electronAPI) {
                await window.electronAPI.showNotification(
                    'Timer Complete',
                    `Timer '${timerName}' has finished!`
                );
            }
        } catch (error) {
            console.warn('Desktop notification failed:', error);
        }
    }

    // Method to restart alerts with new settings
    restartActiveAlerts() {
        const activeNames = Array.from(this.activeAlerts.keys());
        for (const timerName of activeNames) {
            if (this.activeAlerts.has(timerName)) {
                // Stop and restart alert with new settings
                this.stopAlert(timerName);
                setTimeout(() => {
                    this.startAlert(timerName);
                }, 100);
            }
        }
    }
}

// Make AlertManager available globally
window.AlertManager = AlertManager;